runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -1384,6 +1374,17 @@ private void process(
       Object executionKey =
           keyCoder == null ? null : keyCoder.decode(key.newInput(), Coder.Context.OUTER);
 
+      if (workItem.hasHotKeyInfo()) {
+        Windmill.HotKeyInfo hotKeyInfo = workItem.getHotKeyInfo();
+        Duration hotKeyAge = Duration.millis(hotKeyInfo.getHotKeyAgeUsec() / 1000);
+
+        // The MapTask instruction is ordered by dependencies, such that the first element is
+        // always going to be the shuffle task.
+        String stepName = computationState.getMapTask().getInstructions().get(0).getName();
+        Object hotkey = options.isHotKeyLoggingEnabled() ? executionKey : null;
+        hotKeyLogger.logHotKeyDetection(stepName, hotKeyAge, hotkey);
```

Comment:
See my earlier comment on #12984  about assuming that `key == null` is the same thing as not present.

Reply:
Done